File: compliance/python/workflows/callback_approval.py
"""
Callback Approval Compliance Workflow (Python reference implementation).

Callback-based approval workflow: submit request, create callback for
external approval, wait for timeout, check approval result, finalize.

Usage:
    python callback_approval.py > ../tests/fixtures/callback_approval.json
"""

import json


def generate_expected_operations():
    """Generate the operation sequence that this workflow produces.

    The Python durable Lambda SDK records each operation as:
    - type: "step" for ctx.step() calls
    - type: "callback" for ctx.create_callback() calls
    - type: "wait" for ctx.wait() calls

    This workflow performs:
    1. submit_request — step: prepares and submits the approval request
    2. approval — callback: creates callback and waits for external signal
    3. approval_timeout — wait: waits for a timeout period
    4. process_approval — step: processes the approval decision
    5. finalize — step: finalizes the workflow
    """
    return {
        "workflow": "callback_approval",
        "operations": [
            {"type": "step", "name": "submit_request"},
            {"type": "callback", "name": "approval"},
            {"type": "wait", "name": "approval_timeout"},
            {"type": "step", "name": "process_approval"},
            {"type": "step", "name": "finalize"},
        ],
    }


if __name__ == "__main__":
    print(json.dumps(generate_expected_operations(), indent=2))

File: compliance/rust/src/callback_approval.rs
//! Callback Approval compliance workflow.
//!
//! Callback-based approval workflow: submit request → create callback for
//! external approval → wait for timeout → process approval → finalize.
//! Mirrors the Python reference implementation in `compliance/python/workflows/callback_approval.py`.

use durable_lambda_core::context::DurableContext;
use durable_lambda_core::error::DurableError;
use durable_lambda_core::types::CallbackOptions;

/// Execute the callback approval workflow against the given context.
///
/// Operations recorded:
/// 1. `step:submit_request`
/// 2. `callback:approval`
/// 3. `wait:approval_timeout`
/// 4. `step:process_approval`
/// 5. `step:finalize`
pub async fn run(ctx: &mut DurableContext) -> Result<serde_json::Value, DurableError> {
    // Step 1: Submit approval request
    let request: Result<serde_json::Value, String> = ctx
        .step("submit_request", || async {
            Ok(serde_json::json!({"request_id": "REQ-001", "submitted": true}))
        })
        .await?;

    // Step 2: Create callback for external approval
    let handle = ctx
        .create_callback("approval", CallbackOptions::new())
        .await?;
    let approval_result: String = ctx.callback_result(&handle)?;

    // Step 3: Wait for approval timeout
    ctx.wait("approval_timeout", 300).await?;

    // Step 4: Process the approval decision
    let processed: Result<serde_json::Value, String> = ctx
        .step("process_approval", || async {
            Ok(serde_json::json!({"approved": true, "decision": "approved"}))
        })
        .await?;

    // Step 5: Finalize
    let finalized: Result<serde_json::Value, String> = ctx
        .step("finalize", || async {
            Ok(serde_json::json!({"finalized": true, "status": "complete"}))
        })
        .await?;

    Ok(serde_json::json!({
        "request": request.unwrap(),
        "approval": approval_result,
        "processed": processed.unwrap(),
        "finalized": finalized.unwrap(),
    }))
}
